src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Collections/SegmentedArray`1.cs

```diff
@@ -0,0 +1,358 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Runtime.CompilerServices;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.Shared.Collections
+{
+    internal readonly struct SegmentedArray<T> : ICloneable, IList, IStructuralComparable, IStructuralEquatable, IList<T>, IReadOnlyList<T>
+    {
+        private static readonly int s_segmentSize = SegmentedArrayHelper.CalculateSegmentSize(Unsafe.SizeOf<T>());
+        private static readonly int s_segmentShift = SegmentedArrayHelper.CalculateSegmentShift(s_segmentSize);
+        private static readonly int s_offsetMask = SegmentedArrayHelper.CalculateOffsetMask(s_segmentSize);
+
+        private readonly int _length;
+        private readonly T[][] _items;
+
+        public SegmentedArray(int length)
+        {
+            if (length < 0)
+            {
+                throw new ArgumentOutOfRangeException(nameof(length));
+            }
+
+            if (length == 0)
+            {
+                _items = Array.Empty<T[]>();
+                _length = 0;
+            }
+            else
+            {
+                _items = new T[(length + s_segmentSize - 1) >> s_segmentShift][];
+                for (var i = 0; i < _items.Length - 1; i++)
+                {
+                    _items[i] = new T[s_segmentSize];
+                }
+
+                _items[^1] = new T[length & s_offsetMask];
+                _length = length;
+            }
+        }
+
+        private SegmentedArray(int length, T[][] items)
+        {
+            _length = length;
+            _items = items;
+        }
+
+        public bool IsFixedSize => true;
```

Comment:
interesting... so is 'Add' not supported?

Reply:
got it now.

Reply:
Covered already in the PR comment above. Will clarify in the comment for `SegmentedArray<T>` as well.